Failed to apply relocation of type R_X86_64_PC32 to absolute

Assembling the following program with `GNU assembler (GNU Binutils) 2.45.1`

```asm
.text
.globl main
.type main, @function
main:
    callq 0
    ret
.size main, .-main

.text
.globl _start
.type _start, @function
_start:
    call main
    ud2
.size _start, .-_start

.section .note.GNU-stack,"",@progbits
```

and then attempting to link the object file results in an error:

```
wild: error: Failed copying from ./minasm.o (256 (1/0)) to output file
  Caused by:
    Failed to apply relocations in section `.text` of ./minasm.o
    Failed to apply relocation of type R_X86_64_PC32 to absolute at offset 0x1
    Unsupported absolute relocation
```


`lld`, `ld`, `mold` all seem to be OK with this relocation (no warning or error)

On x86_64, AFAIK, the call instruction is always PC-relative, so for a call to an absolute address, I think the assembler doesn't have much choice but to emit a PC-relative relocation.

It should be possible to make wild handle it provided the output file isn't position-independent. For position-independent binaries (or shared objects) there's not really any good solution since the relative offset to an absolute address isn't known at link time. For a static-pie binary, mold issues an error in that case:

```
gcc -static-pie -nostdlib a.s -o a.out -fuse-ld=mold
mold: error: /tmp/cckXomU3.o:(.text): R_X86_64_PC32 relocation at offset 0x1 against symbol `' can not be used; recompile with -fPIC
collect2: error: ld returned 1 exit status
```

That seems reasonable to me. BFD and LLD  both successfully link, but the produced binary won't call the requested address. Instead they call the requested address + the load offset.

```
objdump -dr a.out

a.out:     file format elf64-x86-64


Disassembly of section .text:

0000000000001295 <main>:
    1295:	e8 66 ed ff ff       	call   0 <main-0x1295>
    129a:	c3                   	ret

000000000000129b <_start>:
    129b:	e8 f5 ff ff ff       	call   1295 <main>
    12a0:	0f 0b                	ud2
```

So I think probably we should fix this for non-position-independent executables and possibly see if we can improve the error message a bit for position-independent outputs.

Out of curiosity, why are you calling an absolute address?

If you need to call an absolute address and your output is position-independent, then I think you'd need to load the address into a register, then use that to call instead.

That's great that you're using libwild. If you need any API changes, we'd be happy to discuss.

--- a/libwild/src/elf_writer.rs
+++ b/libwild/src/elf_writer.rs
@@ -2854,9 +2854,22 @@ fn apply_relocation<
 
             return Ok(RelocationModifier::Normal);
         }
+        RelocationKind::Relative if rel.symbol().is_none() => {
+            if layout.symbol_db.output_kind.is_relocatable() {
+                bail!(
+                    "relocation of type {} to absolute address cannot be used in \
+                    position-independent output; recompile with -fPIC",
+                    rel.raw_type()
+                );
+            }
+            let place = section_info.section_address + offset_in_section;
+            let value = (rel.addend() as u64).wrapping_sub(place);
+            A::relocation_from_raw(r_type)?
+                .write_to_buffer(value, &mut out[offset_in_section as usize..])?;
+            return Ok(RelocationModifier::Normal);
+        }
         _ => {}
     }
-
     let (resolution, symbol_index, local_symbol_id) = get_resolution(rel, object_layout, layout)?;
     let flags = layout.flags_for_symbol(local_symbol_id);
     let mut next_modifier = RelocationModifier::Normal;

--- a/wild/tests/sources/elf/abs-reloc/abs-reloc.s
+++ b/wild/tests/sources/elf/abs-reloc/abs-reloc.s
@@ -1,0 +1,17 @@
+// Test that R_X86_64_PC32 to absolute address is handled for non-PIE executables.
+// `call 0` generates R_X86_64_PC32 with no symbol and addend=-4.
+// lld and GNU ld handle this by computing a PC-relative offset to address 0.
+//#Arch:x86_64
+//#LinkArgs:-nostdlib -z noexecstack
+//#Mode:static
+//#RunEnabled:false
+
+.text
+.globl _start
+.type _start, @function
+_start:
+    callq 0
+    ret
+.size _start, .-_start
+
+.section .note.GNU-stack,"",@progbits

